File: 03-sliding_window/00424-Longest_Repeating_Character_Replacement/00424-longest_repeating_character_replacement.cpp
// 424: Longest Repeating Character Replacement
// https://leetcode.com/problems/longest-repeating-character-replacement/

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    // SOLUTION
    int characterReplacement(string s, int k) {
        vector<int> count(26);
        int maxCount = 0;
        int i = 0;
        int j = 0;

        int result = 0;
        
        while (j<s.size()) {
            count[s[j]-'A']++;
            maxCount = max(maxCount, count[s[j]-'A']);
            
            if (j-i+1-maxCount>k) {
                count[s[i]-'A']--;
                i++;
            }

            result = max(result, j-i+1);
            j++;
        }
        
        return result;
    }
};


int main() {
    Solution o;

    // INPUT
    string s = "ABAB";
    int k = 2;

    // OUTPUT
    auto result = o.characterReplacement(s, k);
    cout<<result<<endl;

    return 0;
}

File: 03-sliding_window/00424-Longest_Repeating_Character_Replacement/00424-longest_repeating_character_replacement.py
# 424: Longest Repeating Character Replacement
# https://leetcode.com/problems/longest-repeating-character-replacement/


class Solution:
    # SOLUTION
    def characterReplacement(self, s: str, k: int) -> int:
        count = [0]*26
        maxCount = 0
        i, j = 0, 0

        result = 0

        while j<len(s):
            count[ord(s[j])-ord('A')]+=1
            maxCount = max(maxCount, count[ord(s[j])-ord('A')])

            if j-i+1-maxCount > k:
                count[ord(s[i])-ord('A')]-=1
                i+=1
            result = max(result, j-i+1)
            j+=1
        
        return result




if __name__ == "__main__":
    o = Solution()

    # INPUT
    s: str = "ABAB"
    k: int = 2

    # OUTPUT
    result = o.characterReplacement(s, k)
    print(result)